### DISCUSSION TITLE
Trained a Custom YoloV11 OBB Model with Custom Dataset, But Getting No Detections

### DISCUSSION THREAD

**Discussion by natenberenstein** [ultralytics/ultralytics#18197] [Q&A] [answered]
# Environment
```
python=3.12.7
torch=2.5.1
torchvision=0.20.1
ultralytics=8.3.38
ultralytics-thop-2.0.12
opencv-python=4.10.0.84
nvidia-cuda-cupti-cu12=12.4.127
nvidia-cuda-nvrtc-cu12=12.4.127
nvidia-cuda-runtime-cu12=12.4.127
```

# Dataset
Object detection dataset was created using cvat.  The dataset consists of Keurig images and the labeled objects are the buttons on the coffee machine.  I modified the data.yaml to fit the YoloV11 data formatting.  The dataset has the following structure:
- images/
  - train/*.png
  - val/*.png
- labels
  - train/*.txt
  - val/*.txt
- data.yaml

data.yaml has the following contents
```txt
Train: Train.txt
names:
  0: 4oz
  1: 6oz
  2: 8oz
  3: 10oz
  4: 12oz
  5: strong
  6: iced
  7: hot
  8: add_water
  9: descale
  10: handle
  11: water bin
  12: power
path: .
train: images/train
val: images/val
```

Example of what the labeled contents will have:
```txt
4 0.590707 0.701204 0.687732 0.701204 0.687732 0.756579 0.590707 0.756579
9 0.524006 0.534504 0.633152 0.534504 0.633152 0.575587 0.524006 0.575587
8 0.365464 0.542247 0.508409 0.542247 0.508409 0.588087 0.365464 0.588087
5 0.359874 0.598838 0.511762 0.586760 0.519713 0.633986 0.367825 0.646064
7 0.509094 0.587168 0.660982 0.575090 0.668933 0.622316 0.517045 0.634393
0 0.426112 0.659534 0.511870 0.659534 0.511870 0.704180 0.426112 0.704180
1 0.537865 0.648219 0.611502 0.648219 0.611502 0.698229 0.537865 0.698229
2 0.385390 0.716093 0.473756 0.716093 0.473756 0.772652 0.385390 0.772652
3 0.486745 0.709545 0.576848 0.709545 0.576848 0.763725 0.486745 0.763725
6 0.642680 0.636316 0.712857 0.636316 0.712857 0.685132 0.642680 0.685132
12 0.428351 0.364767 0.510677 0.364767 0.510677 0.417925 0.428351 0.417925
```

# Code
This is the code I used to train my custom YoloV11 OBB model with my custom dataset
```python
import cv2
from ultralytics import YOLO
model = YOLO("yolo11n.yaml")
model.train(data="<path_to_datasets>/datasets/v19/data.yaml", epochs=10, imgsz=640)

model = YOLO("/home/ncbernar/.pyenv/runs/detect/train5/weights/best.pt")
results = model("<path_to_keurig_image>/keurig.png")

for r in results:
    im_array = r.plot()  # plot a BGR numpy array of predictions
    im = cv2.cvtColor(im_array, cv2.COLOR_BGR2RGB)  # convert to RGB for display
    cv2.imshow("Detection Results", im)
    cv2.waitKey(0)  # wait for a key press to close the window

cv2.destroyAllWindows()
```

After results is collected, I get the following message:
```
image 1/1 <path_to_image>/keurig_top_view-Copy1.png: 640x448 (no detections), 34.7ms
Speed: 1.6ms preprocess, 34.7ms inference, 0.4ms postprocess per image at shape (1, 3, 640, 448)
```

To reproduce this issue, please use the following dataset from [here](https://github.com/BeaverBytes/reproducible-issues/tree/main/ultralytics/datasets/v18).

# Problem
As shown above, results tells me that the model was unable to detect any of the objects, even though the labels and images were provided as part of the dataset.  One example of what the labeled image looks like is below:
![image](https://github.com/user-attachments/assets/e1c5d339-0480-4787-8721-f85045c77675)

But after I feed the model an image for inference, I get the following result with no bounding boxes:
![image](https://github.com/user-attachments/assets/78889e64-277f-46d8-be67-2b5a269206a0)

# Discussion
Basically, I expected the model to be able to detect the obejcts, or buttons, in the image.  It acknowledged the dataset and labels, but still no sign of the bounding boxes being shown on the resulting [image.](url)

Any reason for why this happens?  I suspect that my dataset may not be correctly formatted for the OBB model, but it could be a different issue.

**Comment by pderrenger** [comment]
👋 Hello @BeaverBytes, thank you for sharing details about your custom training experience with Ultralytics 🚀! We understand the effort you’ve put into setting up your custom YoloV11 Object Bounding Box (OBB) model, and we’re here to help.

While we review your discussion, here are a few suggestions that might help you identify the issue:

If this is a 🐛 Bug Report, please make sure to provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/), including a representative data sample and code snippet to allow us to replicate the issue on our end. If your provided example includes all necessary information, thank you!

If this is a custom training ❓ Question, it's important to ensure your dataset adheres to YOLO OBB label formatting and that you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Since OBBs require a specific annotation format, double-check that all bounding box corners are correctly defined in your label files.

## Docs and Community Resources

- 📚 For training and inference guidance, visit the [Ultralytics Docs](https://docs.ultralytics.com).
- For OBB-specific tutorials, confirm you’ve reviewed examples pertinent to your model configuration under the [YOLO modes](https://docs.ultralytics.com/modes/).
- Engage with our vibrant community to exchange ideas and solutions:
  - [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions.
  - [Discourse](https://community.ultralytics.com) for detailed questions and solutions.
  - [Subreddit](https://reddit.com/r/Ultralytics) for shared insights and advice.

## Verify Installation and Dependencies

Please ensure you're running the latest version of `ultralytics` along with updated dependencies required for optimal performance on systems like [CUDA](https://developer.nvidia.com/cuda), [PyTorch>=1.8](https://pytorch.org/), and [Python>=3.8](https://www.python.org/). Update your Ultralytics installation with the following:

```bash
pip install -U ultralytics
```

Additionally, confirm that your system and Python packages match the required versions specified in the [requirements file](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

## Environments

Test your training and inference setup on validated environments to ensure consistency:

- Free cloud GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Cloud services like GCP ([Quickstart](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)) and AWS ([Quickstart](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)).
- [Docker Images](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) for isolated environments: <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Training Suggestions and Debugging Checklist

- Verify your `data.yaml` file is properly formatted and paths are correct.
- Ensure your label files strictly use YOLO OBB standards for annotation.
- Review training logs for any anomalies during training that could indicate a potential misconfiguration, overfitting, or underfitting.

## Status

Our Continuous Integration (CI) tests verify the latest commit’s functionality on macOS, Windows, and Ubuntu. Check the project's current [CI status](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule):

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

---

This is an automated response to get you started 😊. An Ultralytics engineer will review your question and provide additional help soon!

**Comment by Y-T-G** [comment, marked as answer]
1. `yolo11n.yaml` trains a normal detection model, not OBB. And start training from "yolo11n-obb.pt" not yaml. https://docs.ultralytics.com/tasks/obb/
2. 10 epochs is not enough. Set it to 100. 
3. You probably don't have enough images either. Follow the guidelines here: https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/

**Reply by natenberenstein** [reply]
Ahh sorry, I meant to put "yolo11n-obb.yaml", not "yolo11n.yaml".  But I will give this a try!  Thanks for the response :)

**Reply by natenberenstein** [reply]
Sir, you are a g.  I'm able to see the labels on my images now thanks to your advice.   Basically, I just transferred the weights from the pretrained model to the yaml file where I want to build a custom model based on my dataset and trained the model with 100 epochs.  

Thanks again for the help!

**Reply by glenn-jocher** [reply]
You're very welcome! Great to hear it's working now. All credit goes to the YOLO community and the Ultralytics team for making these tools so effective. Happy detecting!

**Comment by Shahar-Nadiv** [comment]
i came across another related issue with the labels, when starting the training all epochs precision output is 0.999 - 1.0,
seems like im doing something wrong maybe passing wrong format of lables?

**Reply by glenn-jocher** [reply]
@Shahar-Nadiv precision values of 0.999-1.0 during training typically indicate overfitting or label formatting issues. First, verify your OBB labels follow the correct format: `class_index x1 y1 x2 y2 x3 y3 x4 y4` with normalized coordinates between 0-1. Also check that your train/val split doesn't have duplicate or nearly identical images, and consider if your dataset is large enough - small datasets can lead to perfect memorization rather than learning.
